okhttp/src/main/java/com/squareup/okhttp/Request.java

```diff
@@ -233,8 +233,8 @@ public Builder method(String method, RequestBody body) {
       if (body != null && !HttpMethod.permitsRequestBody(method)) {
         throw new IllegalArgumentException("method " + method + " must not have a request body.");
       }
-      if (body == null && HttpMethod.permitsRequestBody(method)) {
-        body = RequestBody.create(null, Util.EMPTY_BYTE_ARRAY);
+      if (body == null && HttpMethod.requiresRequestBody(method)) {
+        throw new IllegalArgumentException("method " + method + " must have a request body.");
```

Comment:
This would create a breaking change in the API, as now it completely disallows an empty `PUT`, `POST` or `PATCH` by the definition of [`com.squareup.okhttp.internal.http.HttpMethod.requiresRequestBody`](https://github.com/square/okhttp/blob/master/okhttp/src/main/java/com/squareup/okhttp/internal/http/HttpMethod.java#L26-L30).

The big issue I was facing over in https://github.com/square/okhttp/pull/1558 was that I needed to send an empty `PUT`. I personally, as a consumer of your library, don't necessarily think it should disallow a request unless the HTTP spec explicitly states certain request methods must adhere to the existence or lack of existence of a request body, which it does not state AFAIK. 

If I can't make an empty `PUT` request, which I can right now, it's just with some unconventional and unintended approaches, then I'm simply SOL and have to go back to apache lib or something crappy. :(

---

It'd be great if we took a couple steps back and looked into allowing an empty `PUT`, and making this more accessible by having an alias `put()` in the `Request.Builder` (described in https://github.com/square/okhttp/pull/1558#issuecomment-93545013)